### ISSUE TITLE
How to add extended view methods?

### ISSUE THREAD

**Issue by WontSeeMeComin** [mdarby/restful_acl#8]
Hey Matt,

Firstly, great gem. We have a complex Linux-like permission setup for our users so a typical role-based solution just wouldn't work. I really have two questions and didn't know where else to ask them:

1.) Is there a way I can either write extensions for the view helper you've included (= ?allowed{url}) so I can determine if a link is allowed then do alternate content? For example, if a link isn't allowed for that user, I'd like to show something else, say a span or div instead. Since it's just a view helper, I don't get a return value that I can use in a Ruby conditional. Is there a way to do this as-is, and if not where can I go to extend it? I can't locate the helper.

2.) I like the idea of having RESTful methods that can each return their own boolean to determine authorization, but for our app those methods are unlikely to change from resource to resource. To keep it DRY, I tried sticking those model methods in a module and including it, but having those 2 class methods seems to really be hosing up the works. Do you know another way I can have resources all reference the same 5 methods without using a module?

Thanks again for such a great job.

**Comment by WontSeeMeComin** [comment]
I just saw arnekolja's post about #2, I've pinged him as well to see if he came up with a solution

**Comment by mdarby** [comment]
Thanks for the compliments -- I'm glad you find it useful ;)

1) This shouldn't be too difficult. I'd monkey with RestfulAcl::Helper#allowed?. Seems like you could wrap line #8 in a conditional; yield if access.allowed?, return some string otherwise.

2). How did the two class methods mess up? Any meaningful error messages? It might be that "self" is being set to the Module and not the class that includes the module. If that's the case I'd see if the Module class has some way of knowing the parent class that its being included by and use a little metaprogramming to insert the two class methods into it.

**Comment by WontSeeMeComin** [comment]
Thanks for the prompt reply...I know you're not getting paid to support this, and I appreciate it. #1 sounds good, if I know where to find it I can get it figured out. #2 seems to be a commonly Googled thing as it turns out. The issue is that for the life of me I can't get the calling object class name from inside the Module. If I wrap the class methods in a "ClassMethods" interior module, I can get the "allowed?" helper method to work, but my controller :has_permission? method can't find the required model methods. It's probably operator error on my end...was just wondering if there's an alternate way of sharing these model methods without Modules, maybe something a little kinder on class methods. Thanks!

**Comment by mdarby** [comment]
1) If you have it installed as a plugin in your Rails app it should be under plugins/restful_acl/lib/restful_acl/helper.rb. If you plan on hacking RESTful_ACL for your application, I'd definitely install it this way.

2) Have you tried something like this?: http://gist.github.com/313840

**Comment by WontSeeMeComin** [comment]
Your example did the trick. I think the problem is I'm still foggy on the interaction between an included Module and its parent class. I was able to set unique values in the parent class as constants and reference them from the Module so the conditional methods behaved differently based on which resource I needed to authorize. If I come up with an extension on allowed? helper, I'll send it your way in case you'd like to include it. Thanks again.
